### PR TITLE
User can add predefined attrs

### PR DESCRIPTION
**Pull request by devjoaov**
[linear issue](https://linear.app/bleu-llc/issue/BAL-189/user-can-add-predefined-attrs)

https://www.loom.com/share/f4d0d582b2c744769ff946c65edf24c4

### CODE REVIEW

File: packages/next-app/package.json [hunk 1]
```diff
@@ -22,7 +22,9 @@
     "@hookform/resolvers": "^3.0.0",
     "@pinata/sdk": "^2.1.0",
     "@radix-ui/colors": "^0.1.8",
+    "@radix-ui/react-accordion": "^1.1.1",
```
Comment (ribeirojose) [inline packages/next-app/package.json:25]: When I first saw this I thought of using their `Collapsible` instead because I thought  Accordion only allowed for one of the items to be open at once. But it sounds like there's a `type="multiple"` prop you can pass to it! 😄

File: packages/next-app/tailwind.config.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ module.exports = {
         ...blackA,
         ...slateDark,
         ...blueDark,
+        ...violet,
```
Comment (ribeirojose) [inline packages/next-app/tailwind.config.js:13]: any reason to add another color?

File: packages/next-app/src/contexts/PoolMetadataContext.tsx [hunk 3]
```diff
@@ -55,6 +56,10 @@ export enum StatusLabels {
   "Close",
 }
 
+function hasNull(element: PoolMetadataAttribute) {
+  return Object.values(element).includes(null);
+}
```
Comment (ribeirojose) [inline packages/next-app/src/contexts/PoolMetadataContext.tsx:61]: this could be anything, right?
```suggestion
function hasNull(element: Record<unknown, unknown>) {
  return Object.values(element).includes(null);
}
```

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx [hunk 4]
```diff
@@ -219,6 +235,7 @@ export default function MetadataAttributesTable({
           poolId={poolId}
           isOwner={isOwner}
           metadataUpdated={metadataUpdated}
+          isMetadataValid={isMetadataValid}
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:238]: We could probably consume from this context directly in `Actions` right?

File: packages/next-app/src/contexts/PoolMetadataContext.tsx [hunk 5]
```diff
@@ -109,6 +115,10 @@ export function PoolMetadataProvider({ children }: { children: ReactNode }) {
     setMetadataUpdated(!isEqual(metadata, originalMetadata));
   }, [metadata, originalMetadata]);
 
+  useEffect(() => {
+    setIsMetadataValid(!metadata.some(hasNull));
```
Comment (ribeirojose) [inline packages/next-app/src/contexts/PoolMetadataContext.tsx:119]: 🤔 So even if a description is null the metadata is invalid? Not sure I think that's useful. Also, any reason not to use the zod type validation here?

File: packages/next-app/src/components/Checkbox.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+          <CheckIcon />
+        </CheckboxPrimitive.Indicator>
+      </CheckboxPrimitive.Root>
+      <label
+        htmlFor={id}
+        className="pl-[15px] text-[15px] leading-8 text-white"
+      >
+        {label}
+      </label>
+    </div>
```
Comment (ribeirojose) [inline packages/next-app/src/components/Checkbox.tsx:31]: WDYT about changing this to blue instead of violet, and `text-white` to `text-slate12` to conform with the colors we already have?

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx [hunk 7]
```diff
@@ -123,14 +123,29 @@ function Row({
...
+              NEEDS EDIT
+            </span>
+          </div>
+        ) : (
+          <>{data.key}</>
+        )}
+      </Td>
       <Td>{data.typename}</Td>
-      <Td>{data.description}</Td>
+      <Td>{data.description ?? "..."}</Td>
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:139]: Why would we add `...` if this is null?

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx [hunk 8]
```diff
@@ -123,14 +123,29 @@ function Row({
...
-          <>{data.value}</>
-          <AttributeLink data={data} />
-        </div>
+        {data.value ? (
+          <div className="flex justify-between gap-2">
+            <>{data.value}</>
+            <AttributeLink data={data} />
+          </div>
+        ) : (
+          "..."
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:147]: Same question as above

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx [hunk 9]
```diff
@@ -123,14 +123,29 @@ function Row({
...
         )}
       </Td>
-      <Td className="min-w-[10rem]">{data.key}</Td>
+      <Td className="min-w-[5rem] max-w-[10rem]">
+        {!data.value || !data.description ? (
+          <div className="flex justify-between">
+            <>{data.key}</>
+            <span className="inline-flex items-center rounded-md bg-red-200 p-0.5 text-xs font-semibold text-red-600">
+              NEEDS EDIT
+            </span>
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:132]: This looks like a badge, should we use that instead? Also, wdyt about using the radix color here instead of TW's?

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx [hunk 10]
```diff
@@ -123,14 +123,29 @@ function Row({
...
+          </div>
+        ) : (
+          <>{data.key}</>
+        )}
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:136]: perhaps this could be 
```suggestion
          <div className="flex justify-between">
            <>{data.key}</>
            {!data.value || !data.description && (
                  <span className="inline-flex items-center rounded-md bg-red-200 p-0.5 text-xs font-semibold text-red-600">
                    NEEDS EDIT
                  </span>
                </div>
            )}
```

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx [hunk 11]
```diff
@@ -50,17 +53,32 @@ export function Actions({
           </span>
         </Tooltip>
 
-        <Button className="border border-blue-500 bg-gray-900  text-blue-500 hover:bg-gray-800 focus-visible:outline-indigo-500">
-          Import template
-        </Button>
+        <Tooltip
+          content={"You are not the pool owner"}
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx:57]: No need for curly braces here

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx [hunk 12]
```diff
@@ -50,17 +53,32 @@ export function Actions({
...
+          disableTooltip={isOwner}
+        >
+          <span tabIndex={0}>
+            <Dialog
+              title={"Import template"}
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx:62]: No need for curly braces here

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx [hunk 13]
```diff
@@ -50,17 +53,32 @@ export function Actions({
...
+              content={<PredefinedMetadataModal />}
+            >
+              <Button
+                className="border border-blue-500 bg-gray-900  text-blue-500 hover:bg-gray-800 focus-visible:outline-indigo-500 disabled:opacity-40"
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx:66]: This is already refactored in main, but can we use Radix's colors here?

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx [hunk 14]
```diff
@@ -41,7 +44,7 @@ export function Actions({
           <span tabIndex={0}>
             <Dialog title={"Add attribute"} content={<PoolMetadataItemForm />}>
               <Button
-                className="bg-indigo-500 text-white hover:bg-indigo-400 focus-visible:outline-indigo-500 disabled:bg-indigo-400"
+                className="bg-indigo-500 text-white hover:bg-indigo-400 focus-visible:outline-indigo-500 disabled:bg-indigo-400 disabled:opacity-40"
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx:47]: You can probably leverage the updated Button component in `main` to avoid having to define all these classes here.

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+  },
+  {
+    name: "Boosted Pools",
+    options: [
+      { name: "Summary", type: "text" },
+      { name: "Changelog", type: "text" },
+      { name: "Risks", type: "text" },
+    ],
+  },
+];
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:51]: would avoid defining new names for things that already have a naming convention
```suggestion
export const ATTRIBUTES_TEMPLATE: Itemplate[] = [
  {
    name: "Basic Data",
    options: [
      { key: "Name", typename: "text" },
      { key: "Symbol", typename: "text" },
      { key: "Creator", typename: "text" },
      { key: "Date creation", typename: "date" },
      { key: "Website", typename: "url" },
    ],
  },
  {
    name: "Managed Pools",
    options: [
      { key: "Report", typename: "text" },
      { key: "History", typename: "text" },
      { key: "Changelog", typename: "text" },
    ],
  },
  {
    name: "Boosted Pools",
    options: [
      { key: "Summary", typename: "text" },
      { key: "Changelog", typename: "text" },
      { key: "Risks", typename: "text" },
    ],
  },
];
```

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+import { usePoolMetadata } from "#/contexts/PoolMetadataContext";
+import { toSlug } from "#/utils/formatStringCase";
+
+interface Itemplate {
+  name: string;
+  options: {
+    name: string;
+    type: (typeof TypenameEnum.enum)[keyof typeof TypenameEnum.enum];
+  }[];
+}
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:22]: No need to define this interface if you add `as const` after the `ATTRIBUTES_TEMPLATE` definition

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+import {
+  AccordionContent,
+  AccordionItem,
+  AccordionTrigger,
+} from "#/components/Accordion";
+import { Checkbox } from "#/components/Checkbox";
+import { usePoolMetadata } from "#/contexts/PoolMetadataContext";
+import { toSlug } from "#/utils/formatStringCase";
+
+interface Itemplate {
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:16]:
```suggestion
interface ITemplate {
```

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+
+  return (
+    <form onSubmit={onSubmit}>
+      <Accordion.Root
+        className="my-4 w-full"
+        type="single"
+        defaultValue="item-1"
+        collapsible
+      >
+        {ATTRIBUTES_TEMPLATE.map((attribute) => (
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:140]: `#tip` you can destructure elements here to make your code easier to parse
```suggestion
        {ATTRIBUTES_TEMPLATE.map(({ name, options }) => (
```

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+              </div>
+              <div className="ml-2 flex flex-col">
+                {attribute.options.map((option) => (
+                  <div key={option.name} className="flex items-center">
+                    <Checkbox
+                      id={`${attribute.name}-${option.name}`}
+                      checked={checkboxes[attribute.name][option.name]}
+                      onChange={() =>
+                        handleCheckboxChange(attribute.name, option.name)
+                      }
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:166]: Do we really need to track state here? If so why use the form?

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+                    />
+                  </div>
+                ))}
+              </div>
+            </AccordionContent>
+          </AccordionItem>
+        ))}
+      </Accordion.Root>
+
+      <div className="flex w-full items-center justify-end gap-2">
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:177]: `#thought` we could probably have `Dialog.Footer` to this to 🤔

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+        <Dialog.Close asChild>
+          <Button
+            type="button"
+            className="border border-indigo-500 bg-gray-700 text-indigo-400 hover:bg-gray-600 focus-visible:outline-indigo-500"
+          >
+            Cancel
+          </Button>
```
Comment (ribeirojose) [inline packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:184]: Can forgo this classNames if using the new Button from `main`

**Review by ribeirojose — APPROVED** [review]
LGTM, left a few questions and suggestions! Would prioritize merging this soon so we can give the grants folks an update of progress here.

The logic in the modal to add common keys was a bit hard to parse and very verbose, I hope we don't have to come back to debug anything there. 

Saw [this discussion](https://github.com/orgs/react-hook-form/discussions/7729#discussioncomment-2102434) you might be interested in if you want to refactor it in the future.

File: packages/next-app/tailwind.config.js (hunk 2 above)
Comment (devjoaov) [reply packages/next-app/tailwind.config.js:13]: it's the check box "checked" color, but we can change it

File: packages/next-app/src/contexts/PoolMetadataContext.tsx (hunk 3 above)
Comment (devjoaov) [reply packages/next-app/src/contexts/PoolMetadataContext.tsx:61]: yes, I was thinking just about the metadata context

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx (hunk 4 above)
Comment (devjoaov) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:238]: Yes, we could. We can move both `metadataUpdated` and `isMetadataValid` to `Actions`

File: packages/next-app/src/contexts/PoolMetadataContext.tsx (hunk 5 above)
Comment (devjoaov) [reply packages/next-app/src/contexts/PoolMetadataContext.tsx:119]: Since we established earlier that the description and value couldn't be null, I thought it would make sense to check the `description` and `value` values. 

And about the zod, I forgot to use it 😢

File: packages/next-app/src/components/Checkbox.tsx (hunk 6 above)
Comment (devjoaov) [reply packages/next-app/src/components/Checkbox.tsx:31]: Agreed 👍

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx (hunk 7 above)
Comment (devjoaov) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:139]: Just to match with the design

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx (hunk 8 above)
Comment (devjoaov) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:147]: ☝️

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx (hunk 9 above)
Comment (devjoaov) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/index.tsx:132]: I don't know much about the Radix colors but I definitely can to try using it

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx [hunk 22]
```diff
@@ -50,17 +53,32 @@ export function Actions({
...
+          <span tabIndex={0}>
+            <Dialog
+              title={"Import template"}
+              content={<PredefinedMetadataModal />}
+            >
+              <Button
+                className="border border-blue-500 bg-gray-900  text-blue-500 hover:bg-gray-800 focus-visible:outline-indigo-500 disabled:opacity-40"
+                disabled={!isOwner}
+              >
+                Import template
```
[the remark this answers is not in the document]
Comment (devjoaov) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/Actions.tsx:69]: Thinking about it, I believe "add predefined attributes" would be a better option, what do you think?

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx (hunk 15 above)
Comment (devjoaov) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:51]: That makes sense

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx (hunk 19 above)
Comment (devjoaov) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:166]: If I'm not mistaken, we don't have the checkbox state on the form as the checkbox is a button a span

File: packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx (hunk 20 above)
Comment (devjoaov) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:177]: and how we will trigger the submit button to the form?

Comment (ribeirojose) [reply packages/next-app/src/app/metadata/[network]/pool/[poolId]/(components)/MetadataAttributesTable/PredefinedMetadataModal.tsx:177]: same as now, `Dialog.Footer` has children and we'd have the same buttons here
